File: combo2-advanced-agentic-engineering/tiny-agent-csharp/src/TinyAgent.Shared/DotEnv.cs
namespace TinyAgent.Shared;

/// <summary>
/// A 20-line <c>.env</c> reader, so the C# path needs no extra NuGet package
/// where Python uses <c>python-dotenv</c>. GIVEN.
/// </summary>
public static class DotEnv
{
    /// <summary>
    /// Load <c>.env</c> from <paramref name="startDirectory"/> or the nearest
    /// ancestor that has one. Existing environment variables always win.
    /// </summary>
    public static void Load(string? startDirectory = null)
    {
        var dir = new DirectoryInfo(startDirectory ?? Directory.GetCurrentDirectory());

        while (dir is not null)
        {
            var candidate = Path.Combine(dir.FullName, ".env");
            if (File.Exists(candidate))
            {
                Apply(candidate);
                return;
            }
            dir = dir.Parent;
        }
    }

    private static void Apply(string path)
    {
        foreach (var raw in File.ReadAllLines(path))
        {
            var line = raw.Trim();
            if (line.Length == 0 || line.StartsWith('#')) continue;

            var eq = line.IndexOf('=');
            if (eq <= 0) continue;

            var key = line[..eq].Trim();
            var value = line[(eq + 1)..].Trim().Trim('"', '\'');

            // Don't clobber a real environment variable with a file value.
            if (Environment.GetEnvironmentVariable(key) is null)
            {
                Environment.SetEnvironmentVariable(key, value);
            }
        }
    }

    /// <summary>The API key, or null. Callers produce their own error message.</summary>
    public static string? ApiKey() => Environment.GetEnvironmentVariable("GOOGLE_API_KEY");

    /// <summary>The model id, defaulting to the workshop model.</summary>
    public static string Model() =>
        Environment.GetEnvironmentVariable("GEMINI_MODEL") is { Length: > 0 } m
            ? m
            : GeminiClient.DefaultModel;
}

File: combo2-advanced-agentic-engineering/tiny-agent-java/shared/src/main/java/ai/octoco/tinyagent/shared/DotEnv.java
package ai.octoco.tinyagent.shared;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * A small {@code .env} reader, so the Java path needs no extra dependency
 * where Python uses {@code python-dotenv}. GIVEN.
 *
 * <p>Unlike the C# port, this does not call the equivalent of
 * {@code Environment.SetEnvironmentVariable} — the JVM has no public API to
 * mutate the current process's environment block once it has started. Values
 * read from {@code .env} are kept in a private map instead, and {@link #get}
 * checks the real environment first so a real environment variable always
 * wins, then falls back to what {@code .env} provided.
 */
public final class DotEnv {

    private static final Map<String, String> LOADED = new ConcurrentHashMap<>();

    private DotEnv() {}

    /**
     * Load {@code .env} from {@code startDirectory} or the nearest ancestor
     * that has one.
     */
    public static void load(Path startDirectory) {
        Path dir = startDirectory.toAbsolutePath().normalize();
        while (dir != null) {
            Path candidate = dir.resolve(".env");
            if (Files.isRegularFile(candidate)) {
                apply(candidate);
                return;
            }
            dir = dir.getParent();
        }
    }

    private static void apply(Path path) {
        List<String> lines;
        try {
            lines = Files.readAllLines(path);
        } catch (IOException e) {
            return;
        }

        for (String raw : lines) {
            String line = raw.trim();
            if (line.isEmpty() || line.startsWith("#")) continue;

            int eq = line.indexOf('=');
            if (eq <= 0) continue;

            String key = line.substring(0, eq).trim();
            String value = stripQuotes(line.substring(eq + 1).trim());
            LOADED.put(key, value);
        }
    }

    private static String stripQuotes(String value) {
        if (value.length() >= 2) {
            char first = value.charAt(0);
            char last = value.charAt(value.length() - 1);
            if ((first == '"' && last == '"') || (first == '\'' && last == '\'')) {
                return value.substring(1, value.length() - 1);
            }
        }
        return value;
    }

    /** A value from the real environment, or from {@code .env} — real environment wins. */
    public static String get(String key) {
        String fromEnv = System.getenv(key);
        return fromEnv != null ? fromEnv : LOADED.get(key);
    }

    /** The API key, or null. Callers produce their own error message. */
    public static String apiKey() {
        return get("GOOGLE_API_KEY");
    }

    /** The model id, defaulting to the workshop model. */
    public static String model() {
        String m = get("GEMINI_MODEL");
        return (m != null && !m.isBlank()) ? m : GeminiClient.DEFAULT_MODEL;
    }
}
